src/librustc/traits/error_reporting.rs

```diff
@@ -550,10 +550,33 @@ impl<'a, 'gcx, 'tcx> InferCtxt<'a, 'gcx, 'tcx> {
                             // error message, let's display it!
                             err.note(&s);
                         } else {
-                            // If we can't show anything useful, try to find
-                            // similar impls.
-                            let impl_candidates = self.find_similar_impl_candidates(trait_ref);
-                            self.report_similar_impl_candidates(impl_candidates, &mut err);
+                            let binops: FxHashMap<&str, &str> = [
+                                ("std::ops::Add", "+"),
+                                ("std::ops::Sub", "-"),
+                                ("std::ops::Mul", "*"),
+                                ("std::ops::Div", "/"),
+                                ("std::ops::Rem", "%"),
+                                ("std::ops::BitAnd", "&"),
+                                ("std::ops::BitOr", "|"),
+                            ].iter().cloned().collect();
+
+                            // For binops, be explicit on what operation is not supported
+                            let trait_def = format!("{:?}",
+                                                    self.tcx.lookup_trait_def(trait_ref.def_id()));
+                            if let Some(op) = binops.get(&trait_def.as_str()) {
+                                err.help(&format!("`{} {} {:?}` has no implementation",
+                                                  trait_ref.self_ty(),
+                                                  op,
+                                                  trait_ref.substs()[1]));
+                            } else if &trait_def == "std::cmp::PartialEq" {
+                                err.help(&format!("can't compare `{}` against `{:?}`",
```

Comment:
`Kind` doesn't implement `Display` and using `Ty` instead would not work for [cases with `impl`](https://github.com/rust-lang/rust/pull/40565/files#diff-34dc3d265b15de4ce7dbb7e9b848d42cR36).